pkg/proxy/util/service.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/proxy"
+)
+
+// ProxyServiceMap maps a service to its service info.
+type ProxyServiceMap map[proxy.ServicePortName]proxy.ServiceInfo
+
+// endpointsChange stores both the previous and current ProxyServiceMap.
+type serviceChange struct {
+	// previous ProxyServiceMap.
+	previous ProxyServiceMap
+	// current ProxyServiceMap.
+	current ProxyServiceMap
+}
+
+// ServiceChangeMap maps a service to its serviceChange.
+type ServiceChangeMap struct {
+	// lock protects items.
+	lock sync.Mutex
+	// items is a service to its serviceChange map.
+	items map[types.NamespacedName]*serviceChange
+}
+
+func NewServiceChangeMap() ServiceChangeMap {
+	return ServiceChangeMap{
+		items: make(map[types.NamespacedName]*serviceChange),
+	}
+}
+
+// Merge adds other ProxyEndpointsMap's elements to current ProxyEndpointsMap, if exists, update it.
+// It returns all other ProxyEndpointsMap's service ports.
+func (sm *ProxyServiceMap) Merge(other ProxyServiceMap) sets.String {
+	existingPorts := sets.NewString()
+	for svcPortName, info := range other {
+		existingPorts.Insert(svcPortName.Port)
```

Comment:
This is going to produce a set like {"80", "8000", "443"} - it's not clear why this is useful or safe, given that the input is a a Map which could have multiple Services in it.  If I recall, we ensure that the map only has a single Service in it, but that's because it was internal.  Now that it is external, the assumptions have to be reduced, or at least made WAY more obvious.

Reply:
> This is going to produce a set like {"80", "8000", "443"} 

No. `svcPortName.Port` is `string` type instead of `int`.

Reply:
> t's not clear why this is useful or safe, given that the input is a a Map which could have multiple Services in it. If I recall, we ensure that the map only has a single Service in it, but that's because it was internal. Now that it is external, the assumptions have to be reduced, or at least made WAY more obvious.

I Agree - we can change 

`existingPorts.Insert(svcPortName.Port)` to `existingPorts.Insert(svcPortName.String())`, and `svcPortName.String()` is unique across all services.

Reply:
It's fixed now, please take a look.